### ISSUE TITLE
luaeval('vim.list(t)') and luaeval('vim.dict(t)') do not work as documented

### ISSUE THREAD

**Issue by dpelle** [vim/vim#3086]
In  :help lua-vim  I see:
```
	vim.list([arg])		Returns an empty list or, if "arg" is a Lua
				table with numeric keys 1, ..., n (a
				"sequence"), returns a list l such that l[i] =
				arg[i] for i = 1, ..., n (see |List|).
				Non-numeric keys are not used to initialize
				the list. See also |lua-eval| for conversion
				rules. Example: >
					:lua t = {math.pi, false, say = 'hi'}
					:echo luaeval('vim.list(t)')
					:" [3.141593, 0], 'say' is ignored
```

However, it does not seem to work. This is what I get with vim.8.1.113 (and older) with the lua feature enabled:
```
:lua t = {math.pi, false, say = 'hi'}
:echo luaeval('vim.list(t)')
[]
```
I.e.  vim.list(t) gives an empty list, whereas the documentation says that it should return [3.141593, 0].

Similarly, the doc says:
```
	vim.dict([arg])		Returns an empty dictionary or, if "arg" is a
				Lua table, returns a dict d such that d[k] =
				arg[k] for all string keys k in "arg" (see
				|Dictionary|). Number keys are converted to
				strings. Keys that are not strings are not
				used to initialize the dictionary. See also
				|lua-eval| for conversion rules. Example:
					:lua t = {math.pi, false, say = 'hi'}
					:echo luaeval('vim.dict(t)')
					:" {'say': 'hi'}, numeric keys ignored
```
But this is the actual behavior:
```
:lua t = {math.pi, false, say = 'hi'}
:echo luaeval('vim.dict(t)')
{}
```
I.e. vim.dict(t) returns an empty dictionary, whereas the documentation says that it should return  {'say': 'hi'}

Here is the full vim version:
```
:version
VIM - Vi IMproved 8.1 (2018 May 18, compiled Jun 24 2018 20:34:08)
Included patches: 1-113
Compiled by pel@pel-cirrus7
Huge version without GUI.  Features included (+) or not (-):
+acl               +cursorshape       +job               -mouse_sysmouse    +scrollbind        +vertsplit
+arabic            +dialog_con        +jumplist          +mouse_urxvt       +signs             +virtualedit
+autocmd           +diff              +keymap            +mouse_xterm       +smartindent       +visual
-autoservername    +digraphs          +lambda            +multi_byte        +startuptime       +visualextra
-balloon_eval      -dnd               +langmap           +multi_lang        +statusline        +viminfo
+balloon_eval_term -ebcdic            +libcall           -mzscheme          -sun_workshop      +vreplace
-browse            +emacs_tags        +linebreak         +netbeans_intg     +syntax            +wildignore
++builtin_terms    +eval              +lispindent        +num64             +tag_binary        +wildmenu
+byte_offset       +ex_extra          +listcmds          +packages          +tag_old_static    +windows
+channel           +extra_search      +localmap          +path_extra        -tag_any_white     +writebackup
+cindent           +farsi             +lua               -perl              -tcl               +X11
+clientserver      +file_in_path      +menu              +persistent_undo   +termguicolors     +xfontset
+clipboard         +find_in_path      +mksession         +postscript        +terminal          -xim
+cmdline_compl     +float             +modify_fname      +printer           +terminfo          -xpm
+cmdline_hist      +folding           +mouse             +profile           +termresponse      +xsmp_interact
+cmdline_info      -footer            -mouseshape        +python/dyn        +textobjects       +xterm_clipboard
+comments          +fork()            +mouse_dec         +python3/dyn       +timers            -xterm_save
+conceal           +gettext           +mouse_gpm         +quickfix          +title
+cryptv            -hangul_input      -mouse_jsbterm     +reltime           -toolbar
+cscope            +iconv             +mouse_netterm     +rightleft         +user_commands
+cursorbind        +insert_expand     +mouse_sgr         -ruby              +vartabs
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H     -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1   -DEXITFREE -DABOR
T_ON_INTERNAL_ERROR
Linking: gcc   -L/usr/local/lib -Wl,--as-needed -o vim    -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinf
o -lnsl  -lselinux   -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2
```

**Comment by dpelle** [comment]
This old message in Sep 2, 2012 from Luis Carvalho
had a patch which added support for vim.list([arg])
and vim.dict([arg]) with table argument:

https://groups.google.com/forum/#!msg/vim_dev/GufKvP2WQZw/R-2B5I1_i7kJ

The patch was applied to the documentation but somehow
never applied to if_lua.c which explains this issue.
The patch at that link unfortunately no longer applies
cleanly to the most recent vim-8.1.131.

**Comment by brammool** [comment]
Dominique wrote:

> This old message in Sep 2, 2012 from Luis Carvalho
> had a patch which added support for vim.list([arg])
> and vim.dict([arg]) with table argument:
> 
> https://groups.google.com/forum/#!msg/vim_dev/GufKvP2WQZw/R-2B5I1_i7kJ
> 
> The patch was applied to the documentation but somehow
> never applied to if_lua.c which explains this issue.
> The patch at that link unfortunately no longer applies
> cleanly to the most recent vim-8.1.131.

Does it work after patch 8.1.0134?

-- 
You're as much use as a condom machine at the Vatican.
                  -- Rimmer to Holly in Red Dwarf 'Queeg'

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
@brammool asked:

> Does it work after patch 8.1.0134?

* vim.list([arg]) now works. However, not exactly as documented for boolean in the list but it looks like the doc was wrong. Fixed in pull request https://github.com/vim/vim/pull/3130 fixes the documentation of vim.list([arg])

* vim.dict([arg] also behaves differently than documented.  Doc says:
    :lua t = {math.pi, false, say = 'hi'}
    :echo luaeval('vim.dict(t)')
    :" {'say': 'hi'}, numeric keys ignored
  But It actually show:
    {'1': 3.141593, '2': v:false, 'say': 'hi'}   (numeric keys not ignored)

**Comment by k-takata** [comment]
> But It actually show:
> {'1': 3.141593, '2': v:false, 'say': 'hi'} (numeric keys not ignored)

I'm not sure which should be updated, the behavior or the document, but maybe updating the document is better?
I also found that some part of the document exceeds 80 columns.
Here is a patch for that.


```diff
--- a/runtime/doc/if_lua.txt
+++ b/runtime/doc/if_lua.txt
@@ -127,7 +127,8 @@ Vim evaluation and command execution, an
 				rules. Example: >
 					:lua t = {math.pi, false, say = 'hi'}
 					:echo luaeval('vim.list(t)')
-					:" [3.141593, v:false], 'say' is ignored
+					:" [3.141593, v:false], 'say' is
+					:" ignored
 <
 	vim.dict([arg])		Returns an empty dictionary or, if "arg" is a
 				Lua table, returns a dict d such that d[k] =
@@ -138,10 +139,12 @@ Vim evaluation and command execution, an
 				|lua-eval| for conversion rules. Example: >
 					:lua t = {math.pi, false, say = 'hi'}
 					:echo luaeval('vim.dict(t)')
-					:" {'say': 'hi'}, numeric keys ignored
+					:" {'1': 3.141593, '2': v:false,
+					:" 'say': 'hi'}
 <
 	vim.funcref({name})	Returns a Funcref to function {name} (see
-				|Funcref|). It is equivalent to Vim's function().
+				|Funcref|). It is equivalent to Vim's
+				|function()|.
 
 	vim.buffer([arg])	If "arg" is a number, returns buffer with
 				number "arg" in the buffer list or, if "arg"
```

**Comment by k-takata** [comment]
Fixed in the latest runtime files update: fd35811ca528de7612f9571bce20ef4c392fe5f7
